ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh

```diff
@@ -163,19 +163,24 @@ function extract_oc(){
     fi
     url="${route}.apps.ci.l2s4.p1.openshiftapps.com"
 
-    #check tools are ready for download
-    local progress; progress="."
+    #check tools are ready for download, if it's unavailable for half hour, return 
+    local retry=5
     while curl -kLs https://${url}/${TARGET_VERSION} | grep -q "Extracting tools"
     do
-        echo -ne "Tools have not yet extracted at the server, please wait ${progress}\r"
-        progress+="."
+        echo -ne "Tools have not yet extracted at the server, retry...\r"
+        (( retry -= 1 ))
+        if (( retry < 0 )); then return 1; fi
         sleep 300
     done
 
+    retry=5
     echo -e "downloading oc ${TARGET_VERSION} from server ${url}"
-    if ! (curl -kL\# https://${url}/${TARGET_VERSION}/openshift-client-linux-${target_arch}${TARGET_VERSION}.tar.gz | tar -C ${OC_DIR} -xvz); then
-        echo >&2 "Failed to extract oc binary" && return 1
-    fi
+    while ! (curl -kL\# https://${url}/${TARGET_VERSION}/openshift-client-linux-${target_arch}${TARGET_VERSION}.tar.gz | tar -C ${OC_DIR} -xvz); do
```

Comment:
It is possible to use `oc adm release extract` command to get the oc cli instead of `curl`, per my experience, seem like oc extract command is more stable than curl.

Reply:
BTW, curl has the option `--retry` if curl is the only choice.

Reply:
Right. I do find the openshift-release-artifacts.apps.ci.l2s4.p1.openshiftapps.com server is not that stable. I'm currently using ansible image for the upgrade step to have the ansible tool. If i'm looking at the correct [Dockerfile](https://github.com/openshift/openshift-ansible/blob/master/images/installer/Dockerfile), seems like it doesn't have oc. That means if we use ansible, we have to use curl to get oc for the first time we use it. Alternative, we can change to cli image but install ansible in the upgrade step. What's your preference?

Reply:
> I do find the openshift-release-artifacts.apps.ci.l2s4.p1.openshiftapps.com server is not that stable.

Do you mean `oc adm release extract` is not stable?

Reply:
ah no. I meant the server we curl from. It bothers other teams as well. https://issues.redhat.com/browse/OCPCRT-157

Reply:
Correct, this is the way to inject oc binary. Also it's the reason why we do not have oc installed by default.

Reply:
aha, let me try! Thanks.